atc/db/pipeline_lifecycle.go

```diff
@@ -0,0 +1,89 @@
+package db
+
+import (
+	sq "github.com/Masterminds/squirrel"
+	"github.com/concourse/concourse/atc/db/lock"
+)
+
+//go:generate counterfeiter . PipelineLifecycle
+
+type PipelineLifecycle interface {
+	ArchiveAbandonedPipelines() (int, error)
+}
+
+func NewPipelineLifecycle(conn Conn, lockFactory lock.LockFactory) PipelineLifecycle {
+	return &pipelineLifecycle{
+		conn:        conn,
+		lockFactory: lockFactory,
+	}
+}
+
+type pipelineLifecycle struct {
+	conn        Conn
+	lockFactory lock.LockFactory
+}
+
+func (pl *pipelineLifecycle) ArchiveAbandonedPipelines() (int, error) {
+	tx, err := pl.conn.Begin()
+	if err != nil {
+		return 0, err
+	}
+
+	defer Rollback(tx)
+
+	latestSuccessfulBuilds := `WITH successful_builds AS (
+      SELECT DISTINCT ON (job_id) id, job_id
```

Comment:
Hm I think I'm a little worried about this query when it is run against a really large deployment with thousands of builds. I quickly ran just this SELECT query against hush-house and it seemed to have taken around 2 minutes to run. I think this query worries me because it has to select out all of the builds that have a succeeded status, which most builds in the database are succeeded (hopefully haha). There is the results of the explain analyze:

```
Unique  (cost=768784.77..785318.94 rows=842 width=8) (actual time=125106.692..126360.440 rows=3023 loops=1)
  ->  Sort  (cost=768784.77..777051.86 rows=3306834 width=8) (actual time=125106.690..125927.008 rows=3351546 loops=1)
        Sort Key: job_id, id DESC
        Sort Method: external sort  Disk: 58984kB
        ->  Index Only Scan using succeeded_builds_ordering_with_rerun_builds_idx on builds  (cost=0.56..320277.94 rows=3306834 width=8) (actual time=9.970..119191.727 rows=3351546 loops=1)
              Heap Fetches: 666449
Planning time: 0.723 ms
Execution time: 126369.643 ms
```

I think what you're trying to achieve is to grab the latest successful build of each job, so I think in order to have the query not need to go through every successful build of a job, you might be able to do something that is like "for each job, order the successful builds in descending order of id but **LIMIT 1**`. Something similar would be like this query here

https://github.com/concourse/concourse/blob/ba4fcfdcad4f9343bfe4153a7539148dec67be52/atc/db/versions_db.go#L418-L423

But the difference is in that query I'm only looking for one build but here you are looking for multiple.

Reply:
Guess our data set we experimented with was a poor representation of real life, see here: https://github.com/concourse/concourse/issues/5699#issuecomment-654475693

In short, I agree with your concerns and this means we should probably go the denormalization route and bring back this PR: https://github.com/concourse/concourse/pull/5834

Regarding:
> do something that is like "for each job, order the successful builds in descending order of id but LIMIT 1"
How different is that from having `SELECT DISTINCT ON (job_id)` in the query, which also results in one result per job_id?

Reply:
I'm trying to think how we would add `LIMIT 1` to the query, but can't think of a way that wouldn't result in us hitting the build table per child pipeline...

Reply:
Something I was thinking about was something like `EXPLAIN ANALYZE SELECT ( SELECT bu.id FROM builds bu where bu.job_id = j.id AND bu.status='succeeded' ORDER BY bu.id LIMIT 1), j.id FROM jobs j;`. The thing is with the limit 1, it only really works with the correct index. So it would require something like `CREATE INDEX deleteme ON builds (job_id, id DESC) where status = 'succeeded';`. 

I am hesitant on adding another index to the builds table.. I'm wondering if we can use the successful_build_outputs table too? That way we don't need the `WHERE status = successful` but it would still require a new index on the table.

Reply:
Building off your query I made this query. I don't have a data set + schema to test its performance. Running it on hush-house returns zero records as expected:
```sql
SELECT p.id, p.parent_job_id, p.parent_build_id
FROM pipelines p
LEFT JOIN jobs j on j.id  = p.parent_job_id
WHERE
(parent_build_id IS NOT NULL 
 AND archived=false 
 AND p.parent_build_id < (SELECT bu.id FROM builds bu where bu.job_id = p.parent_job_id AND bu.status='succeeded' ORDER BY bu.id LIMIT 1)
)
OR
(parent_build_id IS NOT NULL AND j.id IS NULL)
OR
(j.active=false)
```

Reply:
Was curious about the data set size in hush house:
```
select count(*) from pipelines; 357
select count(*) from jobs; 3700
select count(*) from builds; 4,055,731
```
That's a lot of builds!

Reply:
We decided to cover the first case (which is causing all the performance concerns for the query) in `build.Finish()`. When a build finishes it will archive any pipelines that were previously being set by the job.

All other cases (parent pipeline deleted) are still done in GC by one query that no longer has to hit the `builds` table.